File: 0001-0050/0003/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> pos(128, -1);
        int ans = 0;
        for (int i = 0, j = 0; i < s.length(); i++) {
            j = max(j, pos[s[i]] + 1);
            ans = max(ans, i - j + 1);
            pos[s[i]] = i;
        }
        return ans;
    }
};

struct Example {
    string s;
    int ans;
};

int main() {
    vector<Example> examples = {
        {"abcabcbb", 3},
        {"bbbbb", 1},
        {"pwwkew", 3},
    };
    
    Solution *solve = new Solution();
    for (int i = 0; i < examples.size(); i++) {
        int ans = solve->lengthOfLongestSubstring(examples[i].s);
        if (ans == examples[i].ans) {
            cout << "PASS: CASE " << i << endl;
        } else {
            cout << "FAIL: CASE " << i << endl;
        }
    }
    
    return 0;
}

File: 0001-0050/0003/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        pos = [-1] * 128
        ans = 0
        i, j = 0, 0
        while i < len(s):
            j = max(j, pos[ord(s[i])] + 1)
            ans = max(ans, i - j + 1)
            pos[ord(s[i])] = i
            i += 1
        return ans


class Example:
    def __init__(self, s, ans):
        self.s = s
        self.ans = ans


def main():
    examples = [
        Example("abcabcbb", 3),
        Example("bbbbb", 1),
        Example("pwwkew", 3)
    ]

    solve = Solution()
    for i, example in enumerate(examples):
        ans = solve.lengthOfLongestSubstring(example.s)
        if ans == example.ans:
            print(f"PASS: CASE {i}")
        else:
            print(f"FAIL: CASE {i}")


if __name__ == "__main__":
    main()
